compiler/src/dotty/tools/dotc/reporting/diagnostic/messages.scala

```diff
@@ -1209,20 +1209,29 @@ object messages {
            |""".stripMargin
   }
 
-  case class OverloadedOrRecursiveMethodNeedsResultType(tree: Names.TermName)(implicit ctx: Context)
+  case class OverloadedOrRecursiveMethodNeedsResultType private (termName: String)(implicit ctx: Context)
```

Comment:
My thought was to prevent accidental use that might happen by passing in a string directly; I may not understand all the implications, but even with private it seems I can still extract the string in `ErrorMessagesTests.scala`:

```
      val OverloadedOrRecursiveMethodNeedsResultType(treeName) :: Nil = messages
      assertEquals("i", treeName)
```
Seems to work